runtime/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+	inner: Mode,
+}
+
+impl Executor {
+	/// Executor for existing runtime.
+	///
+	/// Deprecated: Exists only to connect with current JSONRPC implementation.
+	pub fn new(executor: TaskExecutor) -> Self {
+		Executor {
+			inner: Mode::Tokio(executor),
+		}
+	}
+
+	/// Synchronous executor, used mostly for tests.
+	pub fn new_sync() -> Self {
+		Executor {
+			inner: Mode::Sync,
+		}
+	}
+
+	/// Spawns a new thread for each future (use only for tests).
+	pub fn new_thread_per_future() -> Self {
+		Executor {
+			inner: Mode::ThreadPerFuture,
+		}
+	}
+
+	/// Spawn a future to this runtime
+	pub fn spawn<R>(&self, r: R) where
+		R: IntoFuture<Item=(), Error=()> + Send + 'static,
+		R::Future: Send + 'static,
+	{
+		match self.inner {
+			Mode::Tokio(ref executor) => executor.spawn(r.into_future()),
+			Mode::Sync => {
+				let _= r.into_future().wait();
+			},
+			Mode::ThreadPerFuture => {
+				thread::spawn(move || {
+					let _= r.into_future().wait();
+				});
+			},
+		}
+	}
+
+	/// Spawn a new future returned by given closure.
+	pub fn spawn_fn<F, R>(&self, f: F) where
+		F: FnOnce() -> R + Send + 'static,
+		R: IntoFuture<Item=(), Error=()> + Send + 'static,
+		R::Future: Send + 'static,
+	{
+		match self.inner {
+			Mode::Tokio(ref executor) => executor.spawn(future::lazy(f)),
+			Mode::Sync => {
+				let _ = future::lazy(f).wait();
+			},
+			Mode::ThreadPerFuture => {
+				thread::spawn(move || {
+					let _= f().into_future().wait();
+				});
+			},
+		}
+	}
+
+	/// Spawn a new future and wait for it or for a timeout to occur.
+	pub fn spawn_with_timeout<F, R, T>(&self, f: F, duration: Duration, on_timeout: T) where
+		T: FnOnce() -> () + Send + 'static,
+		F: FnOnce() -> R + Send + 'static,
+		R: IntoFuture<Item=(), Error=()> + Send + 'static,
+		R::Future: Send + 'static,
+	{
+		match self.inner {
+			Mode::Tokio(ref executor) => {
+				executor.spawn(timeout(f, duration, on_timeout))
+			},
+			Mode::Sync => {
+				let _ = timeout(f, duration, on_timeout).wait();
+			},
+			Mode::ThreadPerFuture => {
+				thread::spawn(move || {
+					let _ = timeout(f, duration, on_timeout).wait();
+				});
+			},
+		}
+	}
+}
+
+impl<F: Future<Item = (), Error = ()> + Send + 'static> future::Executor<F> for Executor {
+	fn execute(&self, future: F) -> Result<(), future::ExecuteError<F>> {
+		match self.inner {
+			Mode::Tokio(ref executor) => executor.execute(future),
+			Mode::Sync => {
+				let _= future.wait();
+				Ok(())
+			},
+			Mode::ThreadPerFuture => {
+				thread::spawn(move || {
+					let _= future.wait();
+				});
+				Ok(())
+			},
+		}
+	}
+}
+
+/// A handle to a runtime. Dropping the handle will cause runtime to shutdown.
+pub struct RuntimeHandle {
+	close: Option<futures::Complete<()>>,
+	handle: Option<thread::JoinHandle<()>>
+}
+
+impl From<Runtime> for RuntimeHandle {
+	fn from(el: Runtime) -> Self {
+		el.handle
+	}
+}
+
+impl Drop for RuntimeHandle {
+	fn drop(&mut self) {
+		self.close.take().map(|v| v.send(()));
```

Comment:
cannot move out of `*self` which is behind a mutable reference
   --> runtime/src/lib.rs:252:3
    |
252 |         self.close();
    |         ^^^^ move occurs because `*self` has type `RuntimeHandle`, which does not implement the `Copy` trait

Reply:
yepp, David's suggestion is incorrect.

`close` is handle for `oneshot::Sender` not a method and it is dropped by calling `send` on it.

Reply:
yeah, it doesn't work because `close` takes ownership of `self` which `drop` doesn't